### ISSUE TITLE
Type definitions?

### ISSUE THREAD

**Issue by RobRoseKnows** [spencermountain/wtf_wikipedia#280]
Hiya, glad to have found `wtf_wikipedia` as it's making my life easier. I was wondering if adding type definitions for typescript is anywhere on the roadmap. Would make integrating with Typescript applications very nice.

**Comment by RobRoseKnows** [comment]
Currently for example, you can't `import wtf from 'wtf_wikipedia'` in TypeScript, as it has no typings. You need to `const wtf = require('wtf_wikipedia')`.

**Comment by spencermountain** [comment]
hey Robert, yeah, i'd like to add typescript typings, and of course, i'd merge a pr with them.

i've tried a few times to support them, but it was unclear how best to host and maintain them.

The best solution is if the typings could be tested with a `npm run test:types` command, so i can get yelled at when i forget to update them.

If you, or anyone is interested in writing these, there's a [api.json](https://github.com/spencermountain/wtf_wikipedia/blob/master/api/index.json) file with all the public methods, which should help atleast a little

cheers

**Comment by RobRoseKnows** [comment]
Yeah I'm working on type definitions now, not sure how to test them though, I'll look into it. Wish I saw `api.json` earlier though, I've been going off the source code lol. I'm pretty sure `api.json` is wrong about some things though, for example, `json()` returns an Object, not an Array.

**Comment by spencermountain** [comment]
ah, thank you!
yeah, I'm sure it is wrong. please fix it!

Let me know if I can help, at all. Some of the options params especially, are pretty gross.

**Comment by MrXyfir** [comment]
@RobRoseKnows How far along are you? Could you share what you've got so far? For now, I'm using the following, just so TypeScript will quit complaining:

```ts
declare module 'wtf_wikipedia' {
  export function fetch(titles: string | string[]): Promise<any>;
}
```

I plan on filling it out as I go, so anything helps. If I get it all completed I'll submit a pull request here.

**Comment by RobRoseKnows** [comment]
@MrXyfir I *think* I finished everything, though I don't know how correct they are: https://github.com/RobRoseKnows/DefinitelyTyped/tree/master/types/wtf_Wikipedia

When I have a chance I'm going to add tests and submit a PR.

**Comment by MrXyfir** [comment]
Great! I'll plug it into my project and see if I discover any issues with it.

Also that link doesn't work (thanks to the capital W): https://github.com/RobRoseKnows/DefinitelyTyped/tree/master/types/wtf_wikipedia

**Comment by RobRoseKnows** [comment]
@MrXyfit: Didn't really know about optional arguments as I'm new to typescript but yes, anywhere it's `... | null | undefined` should be optional. 

`Section['depth']` is private because the method `indentation()` simply returns `this.depth`.

Go ahead and submit a PR if you can.

**Comment by MrXyfir** [comment]
@RobRoseKnows Ah, my bad. I somehow missed `indentation()`.

I can't figure out the pull request, seems to only let me merge with the original repo and not yours. Maybe you can do it: https://github.com/MrXyfir/DefinitelyTyped/tree/master/types/wtf_wikipedia. If not, just copy and paste from it.

I haven't fixed all of the arguments that should be optional but I think the majority have been updated.

**Comment by RobRoseKnows** [comment]
So I merged in your changes and they look good, except now when I try to write the linting tests, I get an error, I think it has to do with the overloaded methods, but I'm not sure. I may just simplify the types so that there are no methods with identical names and see if that fixes it.

**Comment by MrXyfir** [comment]
Hmm. I think the only overloaded function I added was `fetch()`. You could use ` | ` for the `titles` and the return but that starts to be a pain when you use them in TypeScript (I think just strict mode) because then you have to do something like `const doc = await wtf.fetch('test') as wtf.Document` instead of it just automatically knowing the type. Might be easiest to just remove types for multiple docs altogether and then petition @spencermountain for a `fetchAll([...])` or similar instead.

Let me know if there's anything I can do to help.

**Comment by spencermountain** [comment]
hey! yes, thank you. I agree fetch is too overloaded. 
The intentions, i assure you, were good! ;)
I'll look at cleaning it up in future versions.

**Comment by spencermountain** [comment]
if you do `fetch(any, any, any)`, i can adjust it once the method gets cleaned up. Because it will be a breaking change, it's behind a few other things.

**Comment by MrXyfir** [comment]
@RobRoseKnows Any update on this? Is there anything I can do to help?

I ended up removing the overloaded `fetch()` from my own copy of the types as it was causing me trouble and I didn't have any need to fetch multiple documents. Should be fine to just remove it for now if that's what you're stuck on.

**Comment by RobRoseKnows** [comment]
Yeah sorry. I don't really know how to integrate the typings into the test suite of wtf_wikipedia. Currently it's just a fork of DefinitelyTyped, which has its own test suite, but I don't really know how we can add it to what we already have here.

**Comment by spencermountain** [comment]
hey Rob, feel free to make a pr for these without a test thing. 
then we can all work on it together.
prs welcome
cheers

**Comment by spencermountain** [comment]
hey guys, I copy+pasted Rob's types into `./types` on master in `7.6.0`!

I'm not sure it's correctly implemented, but I just went for it. I've played-around in vscode and it pulls-in tooltips, and seems sane-enough.

Thank you **so much** for your help. I had no idea how much of an undertaking this was. I've been throwing so many weird overloads for several years!

one thing I haven't gotten working is the `npm run test:types` [idea](https://github.com/spencermountain/wtf_wikipedia/blob/master/tests/types/index.ts) - I thought if i ran one token smoke-test with `ts-node ./tests/types/index.ts`, it would catch failed types.

ts-node must be more disciplined than how vscode pulls-in types:
```
TypeError: __1.default is not a function
```
can you guys double-check this, when you have a chance?
thank you!

**Comment by MrXyfir** [comment]
@spencermountain Great!

A quick test on my end and the types seem to be working.

As for the error, I don't think that has anything to do with the TS types. Seems like an import error.

I think the issue is actually your `tape` import. Try it without the `* as`. Log both imports to console. Enabling TypeScript's `esModuleInterop` may help.

**Comment by spencermountain** [comment]
hahah, story of my life.
will keep this open
thanks.

**Comment by ducpt-bili** [comment]
hi @spencermountain , thanks for the greate lib.
I am trying to use your lib on Nestjs project, but it seem like the typescript function is not work when i using the syntax:
- import wtf from 'wtf_wikipedia';
Is there any way to using typescript with your library or any guide i can take? 
Thanks

**Comment by spencermountain** [comment]
hey @ducpt-bili nestjs looks neat , but i don't know anything about how it imports types.
I think we're using best-practice on this side, but If you figure out something we can do, please let me know
cheers

**Comment by lionel-rowe** [comment]
`wtf.fetch`'s typing is especially irritating as `null | X | X[]` is conceptually the same as `X[]`, wherein the null case becomes `[]` and the scalar case becomes `[x]`.

```ts
async function wtfFetch(...args: Parameters<typeof wtf.fetch>): Promise<wtf.Document[]> {
    const result = await wtf.fetch(...args)
    return Array.isArray(result) ? result : result == null ? [] : [result]
}
```

**Comment by spencermountain** [comment]
hey @lionel-rowe yeah, that does seem irritating. Is there anything we can do on our side to improve it? This is way beyond my typescript skill. a PR is welcome
cheers

**Comment by lionel-rowe** [comment]
> hey @lionel-rowe yeah, that does seem irritating. Is there anything we can do on our side to improve it? This is way beyond my typescript skill. a PR is welcome

[edited] I think something like the following would work with zero/minimal changes to runtime logic:

```ts
type FetchResult<T> = T extends unknown[]
  ? Document[]
  : Document | null

type FetchOptions = {
  path?: string
  wiki?: string
  domain?: string
  follow_redirects?: boolean
  lang?: string
  "Api-User-Agent"?: string
}

type FetchCallback<T> = (...args: [err: Error, result: null] | [err: null, result: FetchResult<T>]) => void

declare function fetch<T extends string | number | string[] | number[] | URL>(
  title: T,
  options?: FetchOptions | undefined,
  callback?: FetchCallback<T>
): Promise<FetchResult<T>>
```

It seems like `options.title` is always ignored, as it simply gets overwritten by the first argument `title`, so that can be removed from the types.

Usage:

```ts
await wtf.fetch('Mexico') // Document | null
await wtf.fetch('https://en.wikipedia.org/wiki/Mexico') // Document | null
await wtf.fetch(3966054) // Document | null
await wtf.fetch(['Mexico', 'Colombia']) // Document[]
await wtf.fetch([3966054, 5222]) // Document[]

wtf.fetch('Mexico', undefined, (err, doc) => {
  if (err == null) {
    doc // Document
    err // null
  } else {
    doc // null
    err // Error
  }
})
```

**Comment by spencermountain** [comment]
hey cool! Ya, I support any changes that would improve the ts workflow. PRs welcomed.
Yeah - I think the `options.title` thing was bolted-on for the dump-parsing script, which uses it to lift-up the page title. You're welcome to trim it, if it makes sense.
Happy to include this stuff in the next release, if you were or anyone was inclined to look at it. Thank you
cheers
